session.go

```diff
@@ -382,12 +385,33 @@ func (s *Session) waitForSendErr(hdr header, body []byte, errCh chan error) erro
 		return ErrConnectionWriteTimeout
 	}
 
+	bodyCopy := func() {
+		// In the event of session shutdown or connection write timeout,
+		// we need to prevent `send` from using the body buffer after
+		// returning from this function. If this happens, we attempt to
+		// drain the body channel and store a copy in it, unless it's
+		// already been copied into the buffer managed by `send`.
+		select {
+		case _, ok := <-bodyC:
+			if !ok {
+				return // Body was already copied.
+			}
+			newBody := make([]byte, len(body))
+			copy(newBody, body)
```

Comment:
I think there is still a race here where we drain the channel on line 395, then `send()` selects on line 469 and hits the default case, then we write the copy to the channel on line 401.  You end up not writing the data.

Given that we're in session shutdown or write timeout, do we want to even bother copying the body?  Like, we're going to return an error, so I don't think the sender will have the expectation that the data was actually written.

Reply:
Good catch! This subtle error was introduced by the change in [`a9a7d9e`](https://github.com/coder/yamux/pull/4/commits/a9a7d9ea764c2c12cdf32cf1f24395fca9461827), the default case was not supposed to be present any longer.

The reason we care about copying the body is that we have no "cancel" mechanism for writes where the header has already been written, so we allow the message to pass as-is if it reaches this junction. If we don't write the body a future write could fail due to incomplete packet (header, missing body). The idea is to keep the behavior as close to the original implementation as possible, whilst doing it safely.